### PR TITLE
Fixes soundness bug 18510 by aborting on unwind from safe extern "C" functions only

### LINKED ISSUE #18510: prevent unwinding past FFI boundaries in code generation
It's undefined to unwind past an FFI boundary such as a `pub extern "C" fn`. Code generation should automatically insert a landing pad doing an abort. This will eliminate the class of memory safety errors resulting from unwinding into C from Rust. LLVM will be able to optimize it out if it is being caught and handled explicitly, such as to translate into an error code for C.

EDIT: [Mentoring instructions can be found here.](https://github.com/rust-lang/rust/issues/18510#issuecomment-315789713)

### LINKED ISSUE #52652: Abort instead of unwinding past FFI functions
## More updated description (2021-03-11)

With https://github.com/rust-lang/rust/pull/76570 having landed, Rust now aborts when a panic leaves an `extern "C"` function, so the original soundness issue is fixed. However, there is an attribute to opt-out of this behavior, `#[unwind(allowed)]`. Using this attribute is currently unsound, so it should be fixed, removed, or require some form of `unsafe`.

## Updated Description

This is a tracking issue for switching the compiler to abort the program whenever a Rust program panics across an FFI boundary, notably causing any non-Rust-ABI function to panic/unwind.

This is being implemented in #55982 and some regressions were found in the wild so this is also being repurposed as a tracking issue for any breakage that comes up and a place to discuss the breakage. If you're a crate author and you find your way here, don't hesitate to ask questions!

## Original Description

Doing Rust unwinding into non-Rust stack frames is undefined behavior. This was [fixed](https://github.com/rust-lang/rust/pull/46833) in [1.24.0](https://blog.rust-lang.org/2018/02/15/Rust-1.24.html#other-good-stuff), and then reverted (I think by https://github.com/rust-lang/rust/pull/48380 ?) in [1.24.1](https://blog.rust-lang.org/2018/03/01/Rust-1.24.1.html#do-not-abort-when-unwinding-through-ffi) because of a [regression](https://github.com/rust-lang/rust/issues/48251) that affected rlua. 

The latter blog post said:

> The [solution](https://github.com/rust-lang/rust/pull/48572) here is to generate the abort handler, but in a way that `longjmp` won’t trigger it. It’s not 100% clear if this will make it into Rust 1.25; if the landing is smooth, we may backport, otherwise, this functionality will be back in 1.26.

The link PR has landed, but my understanding is that it does *not* change FFI functions back to aborting on unwind (though it looks like it does fix the issue that affected rlua).

This UB is not mentioned in any later release notes.

This issue has been assigned to @BatmanAoD via [this comment](https://github.com/rust-lang/rust/issues/52652#issuecomment-1027361227).

### PR DESCRIPTION
**Pull request by gnzlbg**
This fixes the soundness bug #18510 from November 2014 which allows safe Rust code to invoke undefined behavior:

```rust
extern "C" fn foo() { panic!() }
fn main() { foo() } // UB: a panic escapes an extern "C" function
```

by implementing the behavior defined in the reference for all `extern "C"` functions, which is to abort when a panic! tries to unwind out of them, for safe extern "C" functions only.

That is, unlike previous attempts at fixing this soundness hole, this PR does not make `unsafe extern "C"` functions abort, since that is not required for soundness, and has been shown to break a lot of code (see #52652 top comment for a brief history of the breakage).

That is, after this PR, it is still possible to invoke undefined behavior by unwinding out of an `unsafe extern "C"` function like here:

```rust
unsafe extern "C" fn foo() { panic!() }
fn main() { unsafe { foo() } } // UB: a panic escapes an unsafe extern "C" function
```

but doing so requires unsafe code, and is an implementation-bug, but not a soundness bug.

There is a PR open (#63909) that removes the `nounwind` attribute from `unsafe extern "C"` functions to try to prevent them from being optimized if they exhibit undefined behavior. 

Note: this PR only closes #18510, which has been closed and and then re-opened as #52652. It was pointed out, that a lot of people do not interpret #52652 to be about a soundness bug, even though the top comment points to 18510, and the first couple of comments in the issue clarify that. It was also pointed out that people would prefer to re-purpose #52652 for discussing future language features instead.

### CODE REVIEW

**Comment by petrochenkov** [comment]
r? @RalfJung 
(Not my area.)

**Comment by RalfJung** [comment]
Still on vacation, sorry.
Cc https://github.com/rust-lang/rust/issues/52652 https://github.com/rust-lang/rust/pull/63909 https://github.com/rust-lang/rfcs/pull/2753

**Comment by RalfJung** [comment]
> Closes #52652.

This seems rather inadequate. We still don't properly insert the shims for all functions then, we in particular we do not do anything about mozjpeg and rlua being UB.

**Comment by RalfJung** [comment]
I think I am of mixed mind about this change. It's a hack, but it does turn some UB into guaranteed abort which is nice, but it also doesn't really help to solve what I consider to be [the main problem with the current unwind situation](https://github.com/rust-lang/rust/pull/63909#issuecomment-529393176).

So, I am not necessarily opposed to landing this if there is general support. But I do think that we should definitely keep https://github.com/rust-lang/rust/issues/52652 open (so the "closes" line should be removed from the PR description).

@gnzlbg assuming your goal here is to "agree on and fix what seems uncontroversial", there is also some other "low-hanging fruit" around unwind, I think? Such as https://github.com/rust-lang/rust/issues/52652 and the fact that `extern "Rust"` FFI *imports* are marked `nounwind`. It might make sense to also try and fix those while the FFI story is still up in the air?

Also, this PR should come with a test (we already have a test for `#[unwind(abort)]`, we should have the same test for a safe attribute-free `extern "C"` fn then).

**Comment by gnzlbg** [comment]
> @gnzlbg assuming your goal here is to "agree on and fix what seems uncontroversial", there is also some other "low-hanging fruit" around unwind, I think? Such as #63883 and the fact that extern "Rust" FFI imports are marked nounwind. It might make sense to also try and fix those while the FFI story is still up in the air?

I think your PR (#63909) fixes most of this other low-hanging fruit correctly. We should do that, maybe without the part that removes `nounwind` from safe `extern "C"` definitions. If there is interest in pursuing this approach, we could organize how to land things next week once you come back from the holidays.

> Also, this PR should come with a test (we already have a test for #[unwind(abort)], we should have the same test for a safe attribute-free extern "C" fn then).

Sure, done.

> we in particular we do not do anything about mozjpeg
> We still don't properly insert the shims for all functions then, 

Well, that's on purpose. Right now the status quo is that we have a soundness bug open on safe Rust but we can't fix it because around 2 crates are relying on undefined behavior. Until now our strategy has been to block the soundness bug solution until we figure out which problems those crates actually need solving, word RFCs, merge them, implement them, and only after these features are stabilized and these crates migrated to use them, only them we can fix the soundness bug. 

This PR fixes the soundness bug without breaking nor fixing those crates. This allow us to close the soundness hole and to solve the problems these crate have without any time pressure nor need to come up with quick temporary features.

> So, I am not necessarily opposed to landing this if there is general support. But I do think that we should definitely keep #52652 open (so the "closes" line should be removed from the PR description).

I've removed the "closes" but if we probably should rename / refocus that issue to reflect what remains to be done. I understood #52652 to be about safe Rust code being able to invoke undefined behavior which is what this PR fixes, but that issue has evolved into how to solve the problems these other crates have because that is believed to be a blocker for fixing the soundness hole.

---

> and rlua being UB.

This is kind off-topic but it has reminded me that this PR might actually trigger a different unrelated bug.

First, AFAICT, rlua currently does not have any undefined behavior of this kind. How could it? It never "unwinds" as in panics or throws exceptions that require cleanup code. Instead, rlua only `longjmp`s. There are some platforms like Windows that implement `longjmp` using asynchronous exceptions/SEH, but IIUC [it is correct to unwind out of LLVM `nounwind` functions using those mechanisms](https://llvm.org/docs/LangRef.html#function-attributes) (emphasis mine):

> nounwind: This function attribute indicates that the function never raises an exception. If the function does raise an exception, its runtime behavior is undefined. However, **functions marked nounwind may still trap or generate asynchronous exceptions**. Exception handling schemes that are recognized by LLVM to handle asynchronous exceptions, **such as SEH, will still provide their implementation defined semantics**.

Clang has a correctness [bug](https://bugs.llvm.org/show_bug.cgi?id=36508) in which trying to `longjmp` out of a `noexcept` function on Windows MSVC targets incorrectly aborts the process. We should determine whether rustc has the same bug, because if it has, we are going to run into it sooner or later the moment we start aborting on these functions. 

@kyren do the rlua bindings use any "safe" `extern "C"` callbacks ? If so, then rlua might be affected by such a bug and we should probably fix it before landing this, but AFAICT all the callbacks are `unsafe extern "C"` functions, so they should just work (and it is correct that they are `nounwind`).

---

Does anyone has any idea about what is causing the build errors?

File: src/librustc_mir/build/mod.rs [hunk 1]
```diff
@@ -491,6 +491,7 @@ fn should_abort_on_panic(tcx: TyCtxt<'_>, fn_def_id: DefId, abi: Abi) -> bool {
 
     // Validate `#[unwind]` syntax regardless of platform-specific panic strategy
     let attrs = &tcx.get_attrs(fn_def_id);
+    let unsafety = &tcx.fn_sig(fn_def_id).skip_binder().unsafety.clone();
```
Comment (gnzlbg) [inline src/librustc_mir/build/mod.rs]: @eddyb I don't know if this is the best way to do this. I'm not sure what the binder is for and just ended up playing type tetris here.

**Comment by joshtriplett** [comment]
As @RalfJung mentioned in one of the other issues: I don't believe that we should make code generation depend on the presence or absence of `unsafe`.

This is also missing an explanation of the advantages of this proposal over `#[unwind]`. In particular, assuming that `#[unwind]` goes through in a reasonably timely fashion, and that in the meantime we avoid emitting `nounwind`, this seems like an unnecessary additional complication to introduce in the interim.

**Comment by RalfJung** [comment]
> I've removed the "closes" but if we merge this we probably should rename / refocus that issue to reflect what remains to be done. I understood #52652 to be about safe Rust code being able to invoke undefined behavior which is what this PR fixes, but that issue has evolved into how to solve the problems these other crates have because that is believed to be a blocker for fixing the soundness hole.

Then you understanding widely differs from mine. I was not even aware of the soundness bug until you raised it, which was after this recent burst of activity on this issue.
For me, the bug is "there's something programmers want to do in Rust but they cannot". The fact that this can cause UB in safe code is just a sideshow. Still worth fixing, but certainly not the main issue at hand.

**Comment by gnzlbg** [comment]
@RalfJung 

>  I was not even aware of the soundness bug until you raised it, which was after this recent burst of activity on this issue.

You might be confusing this with the recently reported miscompilation due to this because the soundness bug was discovered at least in 2017, with the first attempted fix landing in [December 2017](https://github.com/rust-lang/rust/pull/46833), and people have been trying to fix it ever since (the first comment in the issue explains this).

---

@joshtriplett 

> This is also missing an explanation of the advantages of this proposal over #[unwind]

This isn't a proposal over `#[unwind]`, this is a fix for the soundness bug being tracked in #52652 - `#[unwind]` can still happen and everything here is forward compatible with that. 

> In particular, assuming that #[unwind] goes through in a reasonably timely fashion, and that in the meantime we avoid emitting nounwind, this seems like an unnecessary additional complication to introduce in the interim.

Even if we were to assume that #[unwind]` would go through in a reasonably timely fashion, we don't have to wait till that happens to fix the soundness bug. 

> and that in the meantime we avoid emitting nounwind, 

First, it isn't clear whether we can avoid emitting nounwind yet as @alexcrichton raised [here](https://github.com/rust-lang/rust/pull/63909#pullrequestreview-280424957). 

> this seems like an unnecessary additional complication to introduce in the interim.

The unnecessary additional complication is avoid emitting `nounwind` in the interim, since emitting `nounwind` and `abort`ing is what we want to do long term, even if `#[unwind]` is merged, and what this PR does.

**Comment by gnzlbg** [comment]
In other words, this PR implements the solution that @rust-lang/lang agreed is the right thing to do for Rust's safe `extern "C"` functions. RFC's like https://github.com/rust-lang/rfcs/pull/2699 (unwind-through-FFI) or  https://github.com/rust-lang/rfcs/pull/2753 (simple_unwind_attribute) have to be (and are) compatible with this.

**Comment by RalfJung** [comment]
> You might be confusing this with the recently reported miscompilation due to the soundness bug, because the soundness bug was discovered at least in 2017, with the first attempted fix landing in December 2017, and people have been trying to fix it ever since. The first comment in the #52652 explains this, and developing new language features to allow programs to unwind through FFI has been discussed pretty much since the start, but it only became "time critical" on March 2019 when it was decided by T-Lang that a fix to the soundness bug shall break no crates.

All of that discussion was about turning UB into a defined abort, and how that broke stuff, but AFAIK never was it mentioned that it is possible to cause this UB from safe code only. There was never a *soundness* bug until [14 days ago](https://github.com/rust-lang/rust/issues/63943).

**Comment by RalfJung** [comment]
>  this is a fix for the soundness bug being tracked in #52652 - #[unwind] can still happen and everything here is forward compatible with that.

See https://github.com/rust-lang/rust/issues/52652#issuecomment-407794075 for a comment explaining that that bug is *not* a soundness bug in the usual sense of the word. Emphasis mine:

> It is indeed **not rustc or std that is unsound**, but the current situation is that it is very difficult to write sound FFI code.

You are misreading history here.

> T-Lang agreed that the right fix for this is to both abort and emit nounwind, and that's what this PR does. 

T-lang agreed that the right fix is to do this *for all `extern "C"`* (except when opting out with an attribute). That is *not* what this PR does.
This PR replaces "opt-out with an attribute" by "opt-out by adding `unsafe`". That's a cute hack. It is by no means a proper fix, and also not fixing any of the issues or implementing any of the proposals that have been floating around, *except for* https://github.com/rust-lang/rust/issues/63943.

**Comment by gnzlbg** [comment]
> All of that discussion was about turning UB into a defined abort, and how that broke stuff, but AFAIK never was it mentioned that it is possible to cause this UB from safe code only. 

That's the only reason #52652 is tagged with I-Unsound:

```rust
extern "C" fn foo() { panic() } // Safe Rust invokes UB
```

> There was never a soundness bug until 14 days ago.

That's just a miscompilation caused by that unsoundness. 

> See #52652 (comment) for a comment explaining that that bug is not a soundness bug in the usual sense of the word. 

See these comments refuting that comment, starting at this one: https://github.com/rust-lang/rust/issues/52652#issuecomment-407907420 and the couple that follow showing UB in safe Rust.

**Comment by gnzlbg** [comment]
> T-lang agreed that the right fix is to do this for all extern "C" (except when opting out with an attribute). 

Sure, as I mention, this PR implements the part of that required to fix the soundness hole, leaving the rest for later. 

> This PR replaces "opt-out with an attribute" by "opt-out by adding unsafe". 

On stable Rust, there is no attribute, so it is unclear to me how this PR could replace that with anything. On nightly Rust, there is an attribute, but this PR does not replace it.

**Comment by RalfJung** [comment]
> Particularly this comment: #52652 (comment) which shows the same UB that leads to the miscompilation I filled two weeks ago.

Fair, I had missed that. (It's not "the first comment though" as you had claimed.)

But #52652 was marked I-unsound before that already. It is not, primarily, a soundness bug. It primarily is about letting people write FFI code like mozjpeg does. The issue description does not even mention (un)soundness.

> On stable Rust, there is no attribute, so it is unclear to me how this PR could replace that with anything. On nightly Rust, there is an attribute, but this PR does not replace it.

It's conceptually replacing it, as in, it is doing the same thing.
Having both the attribute and `unsafe` available as a switch here doesn't even make much sense (and if I understand your comments, you also do not intend it as long-term solution).

**Comment by gnzlbg** [comment]
> (It's not "the first comment though" as you had claimed.)
> But #52652 was marked I-unsound before that already. 

It's linked in the first comment (in the original comment part, not the updated part at the top), which links to the PR that fixed this for the first time, where this was discussed, and the announcement of Rust 1.24.0 which mentions that we fix a case where you could invoke undefined behavior in safe Rust: https://blog.rust-lang.org/2018/02/15/Rust-1.24.html#other-good-stuff

The reason that this issue was marked with I-Unsound initially is that the first and only report of this that one can find is the PR that actually fixed it for the first time. If there were any other issues filled before that, I can't find them. That PR was merged, stuff broke, reverted, and this issue was filled at some point after.. so I guess this is another reason to not use PRs as bug reports =/ 

> It's conceptually replacing it, as in, it is doing the same thing.
> Having both the attribute and unsafe available as a switch here doesn't even make much sense (and if I understand your comments, you also do not intend it as long-term solution).

Right now, even if we stop emitting `nounwind`, both safe and unsafe extern fns that panic are UB. It's UB that we might not exploit, but it is still UB. That is, right now, both the safe and unsafe extern "C" function cases are unsafe.

You might see fixing the safe function case as making the unsafe case the opt out of safety, but that's already the case. 

Sure, until a proper solution lands, this might make users that are invoking undefined behavior in safe Rust code change their code to "only" invoke undefined behavior in unsafe Rust code. Not a great improvement, but IMO an improvement nevertheless.

**Comment by kyren** [comment]
> @kyren do the rlua bindings use any "safe" extern "C" callbacks ? If so, then rlua might be affected by such a bug and we should probably fix it before landing this, but AFAICT all the callbacks are unsafe extern "C" functions, so they should just work (and it is correct that they are nounwind).

No, absolutely all rust functions marked `extern "C"` in rlua are unsafe, so this would not break rlua.

**Comment by joshtriplett** [comment]
@gnzlbg I don't recall a distinction between safe and unsafe functions ever coming up in the language team discussions, and it certainly wasn't in the conclusion from the lang team meeting. The conclusion from the language team meeting is that the state we'd like to *get* to is a consistent one in which *all* functions (safe or unsafe) not marked as unwinding are both marked `nounwind` for optimization by default, and that both abort on attempted unwind. (Thanks to @RalfJung for summarizing that point while I was in the process of writing this comment.) The language team also agreed that our current inconsistent state (not aborting but still adding `nounwind`) is not right; we talked about merging the PR to stop adding `nounwind` *in the interim* to address the current UB as soon as possible, as the RFC for `#[unwind]` seems to keep getting derailed. (Ralf has said the same thing, and I would like to just merge Ralf's PR.)

> Right now, even if we stop emitting nounwind, both safe and unsafe extern fns that panic are UB.

Are you referring to the lack of specification for how Rust unwinding interacts with other languages? If so: RFC 2753 is trying to address that through some minimal additions of specification/definition.

In the meantime, I personally am inclined to merge Ralf's PR at https://github.com/rust-lang/rust/pull/63909 to fix potential miscompilation, and that was the lang team consensus as well, though I'm going to go confirm that again. With that PR merged, we shouldn't generate any further miscompilation. We would miss some optimization opportunities, but we can get those optimization opportunities back once we introduce `#[unwind]` (or some other means for developers to make rust stop emitting `nounwind` on `extern "C"` functions that might unwind). And yes, actually *using* the unwind behavior across languages will still be undefined, but in a well-understood way (namely that you need to use the right compiler options), and at least it'll remain functional while work continues on defining its behavior.

*This* PR, on the other hand, introduces some additional semantics to `unsafe`, without an RFC, and without any documentation. Like @RalfJung, I don't think this is a good idea.

**Comment by gnzlbg** [comment]
> Are you referring to the lack of specification for how Rust unwinding interacts with other languages?

Per the Rust language reference, `extern "C"` function are not allowed to unwind, and therefore, it is ok to emit `nounwind` for them. 

> we talked about merging the PR to stop adding nounwind in the interim to address the current UB as soon as possible,

Removing `nounwind` does not remove the undefined behavior nor the soundness hole in the language. It disables an optimization that exploits it, but that's it, if an `extern "C"` function unwinds, the behavior is still undefined according to the language reference. 

>  And yes, actually using the unwind behavior across languages will still be undefined

Not only across languages. The behavior of this safe Rust program is undefined:

```rust
extern "C" fn foo() { panic!() } 
fn main() { foo() }
```

independently of whether we emit `nounwind` or not.

> This PR, on the other hand, introduces some additional semantics to unsafe

Which semantics? This PR does not allow unsafe code to do anything that it cannot already do without this PR independently of whether @RalfJung 's PR is merged or not. This PR does however limit what safe code can do, in particular, it does not let safe Rust examples like the above have undefined behavior.

**Comment by RalfJung** [comment]
> It's linked in the first comment (in the original comment part, not the updated part at the top), which links to the PR that fixed this for the first time, where this was discussed, and the announcement of Rust 1.24.0 which mentions that we fix a case where you could invoke undefined behavior in safe Rust: https://blog.rust-lang.org/2018/02/15/Rust-1.24.html#other-good-stuff

None of this talks about being able to trigger UB in safe code. It's all about having UB on an FFI boundary, so there's unsafe code involved.

AFAIK https://github.com/rust-lang/rust/issues/52652#issuecomment-408123653 is the first time it was shown that safe Rust can run into this. And that comment (somewhat surprisingly I have to say) did not get much of a reaction. The discussion continued to be solely about the FFI case.

The only one talking about the soundness issue since then is you, AFAIK. I wasn't even aware I was fixing a soundness issue when writing https://github.com/rust-lang/rust/pull/63909.

> The reason that this issue was marked with I-Unsound initially is that the first and only report of this that one can find is the PR that actually fixed it for the first time. 

That is incorrect, I quoted above the reason it is marked I-Unsound:

> It is indeed not rustc or std that is unsound, but the current situation is that it is very difficult to write sound FFI code.

(from https://github.com/rust-lang/rust/issues/52652#issuecomment-407794075, which is before the soundness issue was even raised for the first time)

---

Not that it matters very much what happened back then, though... the matter right now is that we have two PRs fixing the soundness issue, this one here and https://github.com/rust-lang/rust/pull/63909.

This one here has the advantage of minimizing the possible performance impact, but OTOH it makes `unsafe` relevant for codegen which is very odd at best.

The other one goes further, risking a bigger perf/codesize impact, but OTOH it also removes "de facto" (but not "de jure") UB from mozjpeg and basically just completes what we started but didn't finish in https://github.com/rust-lang/rust/pull/62603. It is true that this PR only "de facto" fixes the soundness hole @gnzlbg is worried about, but everything about this situation is temporary so that does not seem like a big problem.

---

> Per the Rust language reference, extern "C" function are not allowed to unwind, and therefore, it is ok to emit nounwind for them.

I think @gnzlbg is talking about "de jure" UB here, while @joshtriplett was asking about "de facto" UB. As in, even with my PR, mozjpeg is still UB under the Rust spec, and only an RFC can fix that. But, with my PR, we reliably generate UB-free LLVM IR for mozjpeg.

**Comment by gnzlbg** [comment]
@RalfJung See: https://github.com/rust-lang/rust/issues/18510 which is the original soundness bug report, which does contain the UB in safe Rust down in the comments. That issue was closed when the first PR with a fix for this landed, but when that failed, #52652 was reopened.

> The other one goes further, risking a bigger perf/codesize impact, but OTOH it also removes "de facto" (but not "de jure") UB from mozjpeg and basically just completes what we started but didn't finish in #62603.

FWIW I think we should do both. Remove "de jure" UB from safe Rust code with this PR, and remove "de facto" UB from unsafe Rust code with @RalfJung PR. That's as good as it can get without an RFC, it fixes the soundness hole in safe Rust, and it allows unsafe Rust code to not be "mis"-optimized until an RFC with a better solution lands.

**Comment by gnzlbg** [comment]
So I've updated the top comment to make it clear which problem this solves.

**Comment by Centril** [comment]
Having read the discussion, this PR seems reasonable to land as a temporary crutch so that the soundness hole, which I regard as the primary bug, is resolved.

**Comment by RalfJung** [comment]
> See: #18510 which is the original soundness bug report, which does contain the UB in safe Rust down in the comments. That issue was closed when the first PR with a fix for this landed, but when that failed, #52652 was reopened.

Again that bug was not originally about soundness within Rust, it was all about FFI. Both https://github.com/rust-lang/rust/issues/18510 and https://github.com/rust-lang/rust/issues/52652 are (in their OP description) specifically about FFI. But you are right insofar that the Rust soundness issue was [mentioned in there](https://github.com/rust-lang/rust/issues/18510#issuecomment-333058944).

Still, https://github.com/rust-lang/rust/issues/63943 I think is the first issue that is specifically about the soundness issue fixed here. I think this PR closes https://github.com/rust-lang/rust/issues/63943, does it not? Could you add that to the PR description? Also I think that description mischaracterizes what https://github.com/rust-lang/rust/issues/18510 is about, but whatever.

> by implementing the behavior defined in the reference which requires safe extern "C" functions to abort when a panic! tries to unwind out of them.

This makes it sound like the reference says something specifically about safe extern "C" functions. It does not.

> Having read the discussion, this PR seems reasonable to land as a temporary crutch so that the soundness hole, which I regard as the primary bug, is resolved.

The primary bug in what context? Note that in the lang team discussions, at least the one where I was present, the soundness hole was not even mentioned. There is a whole lot more to this unwinding discussion than this soundness hole.

IMO https://github.com/rust-lang/rust/pull/63909 is the "better" clutch to fix the soundness issue (at last to fix the de-facto UB), and it does not require making codegen depend on the `unsafe` keyword -- a big fat warning sign IMO, and a very dangerous precedent.

**Comment by Centril** [comment]
> The primary bug in what context? Note that in the lang team discussions, at least the one where I was present, the soundness hole was not even mentioned. There is a whole lot more to this unwinding discussion than this soundness hole.

The lang team discussions started as a result of FCP-merge on https://github.com/rust-lang/rust/issues/58794#issuecomment-471281240. I proposed FCP so that the UB in safe code would be resolved. The unwinding aspect took over once that became the blocker for closing the soundness hole.

> a big fat warning sign IMO, and a very dangerous precedent.

I don't disagree and would definitely not like this PR on a permanent basis either. But when it comes down to it, I find that closing the soundness without inviting more reliance on spec-UB is more important.

**Comment by gnzlbg** [comment]
@RalfJung  You are mixing things up.

The Rust language does not allow `extern "C"` functions to unwind. 

The soundness bug is that it is possible to write safe Rust code that causes an `extern "C"` function to unwind and therefore it allows safe Rust to exhibit undefined behavior. This was reported in #18510.

We use the `nounwind` attribute on `extern "C"` functions. Therefore, if an `extern "C"` function unwinds, the behavior is not only undefined in Rust, but also in LLVM-IR. Such undefined behavior in LLVM was reported in #18510 as well.

You [reported](https://github.com/rust-lang/rust/pull/63909#issuecomment-524830388) that you did not known of any actual code for which LLVM optimized its IR by using `nounwind`. I searched for a report of such a case and could not find any, so I filled #63943 to report that this issue was not only theoretical, but also practical.

> I think this PR closes #63943, does it not? 

This PR closes the soundness hole, and therefore when the optimization shown in #63943 happens in safe Rust, it is always correct. However, after this PR that optimization can still happen in unsafe Rust code, so the problem reported in #63943 is still there (just replace `fn` with `unsafe fn`). 

> This makes it sound like the reference says something specifically about safe extern "C" functions. It does not.

I've clarified that to:

> by implementing the behavior defined in the reference for all extern "C" functions, which is to abort when a panic! tries to unwind out of them, only for safe extern "C" functions.

**Comment by RalfJung** [comment]
> The lang team discussions started as a result of FCP-merge on #58794 (comment). I proposed FCP so that the UB in safe code would be resolved. The unwinding aspect took over once that became the blocker for closing the soundness hole.

I see. I was not aware the lang team talked about the soundness issue back then.

Sorry for my misunderstandings here.

> However, after this PR that optimization can still happen in unsafe Rust code, so the problem reported in #63943 is still there (just replace fn with unsafe fn).

But here you are contradicting yourself... "The Rust language does not allow extern "C" functions to unwind", and hence there is no miscompilation in #63943. We can only call this a miscompilation because it is entirely safe code.

**Comment by gnzlbg** [comment]
>  "The Rust language does not allow extern "C" functions to unwind", and hence there is no miscompilation in #63943. 

@RalfJung you are completely right: #63943 is not a miscompilation.

> We can only call this a miscompilation because it is entirely safe code.

I consider the soundness bug and the ""miscompilation"" to be separate issues, but the actual reason I called #63943 a miscompilation is because this completely correct optimization """breaks""" code that has undefined behavior. 

The lang team has set as a hard constraint that solutions to the soundness bug shall not break code that exhibits undefined behavior. This is the new normal. I'm open to renaming the issue to something else that makes things clearer (e.g. "unsound optimization for unsound code" or something like that).

**Comment by joshtriplett** [comment]
I feel that allowing `unsafe` to affect code generation is not something we should do. Let's work to fix this for *all* functions.

**Comment by gnzlbg** [comment]
I agree that we should fix this for all functions, and I also agree that having different code generation between safe and unsafe Rust functions is bad, but I also think that unnecessarily exposing safe Rust code to undefined behavior is way way worse than having slightly different codegen between safe and unsafe Rust functions.

We make no guarantees about codegen, so I personally find the codegen argument very weak.

**Comment by kyren** [comment]
@BatmanAoD has helped me investigate the current situation, and just as a clarification since rlua keeps coming up in these issues, it appears rlua is not actually currently affected by the abort on unwind through extern "C" behavior at all!

I was a bit unsure of the current state of things after the 1.24.1 reversion due to https://github.com/rust-lang/rust/issues/48251, but it *appears* that since https://github.com/rust-lang/rust/pull/48572 is merged that rlua no longer affected by rustc's unwind abort behavior.  This was tested by running rlua tests on windows with the MSVC toolchain using the nightly compiler version 2019-09-20 97e58c0d3 and observing no failures due to improper aborts.  @BatmanAoD informed me that this version of the compiler should contain the new unwind abort behavior.

Sorry for the confusion, I wasn't quite sure of the full resolution following #48251 until now.

**Comment by gnzlbg** [comment]
> This was tested by running rlua tests on windows with the MSVC toolchain using the nightly compiler version 2019-09-20 97e58c0 and observing no failures due to improper aborts.

That compiler does not abort when `extern "C"` functions unwind; you'll probably need to use a nightly from before https://github.com/rust-lang/rust/pull/62603 was merged.

**Comment by kyren** [comment]
Okay, I tried the nightly version 2019-08-24 eeba189cf and also observed the same behavior: that all rlua tests pass.  That compiler should be from a few days before #62603 was merged, correct?

**Comment by gnzlbg** [comment]
> That compiler should be from a few days before #62603 was merged, correct?

Yes, that PR was merged on the 2019-08-26, so that looks correct to me. 

I think that this was the PR that fixed the rlua case with abort enabled: https://github.com/rust-lang/rust/pull/48572

File: src/librustc_mir/build/mod.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_mir/build/mod.rs]: Just... call `.unsafety()`? There should be a method for every component of `ty::FnSig`, on the binder-wrapped version.

**Comment by JohnCSimon** [comment]
Ping from triage
@RalfJung It seems this PR is ready for merging but has sat idle for 10 days.
CC: @gnzlbg @kyren 

Thanks

**Comment by joshtriplett** [comment]
On October 5, 2019 10:17:46 PM GMT+02:00, John Simon <notifications@github.com> wrote:
>Ping from triage
>@RalfJung It seems this PR is ready for merging but has sat idle for 10
>days.
>CC: @gnzlbg @kyren 
>
>Thanks

This PR is not ready for merging, as its correctness is disputed.

**Comment by hdhoang** [comment]
from triaging, this PR has been idle for 1 month. Could the @rust-lang/lang team decide on this please? thanks!

**Comment by BatmanAoD** [comment]
To refresh everyone's memory, there are three proposals, including this one, for fixing the soundness hole *immediately*:

* [Always `abort` on unwind at FFI boundaries](https://github.com/rust-lang/rust/issues/52652) - this has been stabilized and reverted multiple times. There have been claims that some projects (such as mozjpeg) that currently work on stable Rust are broken by this change (but would not be broken on nightly, which provides `#[unwind(allowed)]` to opt out of the `abort`). There are two objections to this claim (paraphrased here without endorsement):
  * In practice, it's not clear that this part of mozjpeg actually works on stable Rust: https://users.rust-lang.org/t/porting-mozjpeg-to-rust-10k-gig/28899/13?u=batmanaod
  * Even if mozjpeg and/or other projects do appear to behave correctly, the behavior of unwinding across FFI boundaries is undefined in Rust, so they are relying on UB and therefore broken already.
* This change, which modifies codegen depending on whether `unsafe` is used (see discussion above).
* [Remove `nounwind` from codegen for `extern` function definitions](https://users.rust-lang.org/t/porting-mozjpeg-to-rust-10k-gig/28899/13?u=batmanaod). This was closed with the assumption that a mechanism for opting out of `nounwind` would be quickly stabilized, but the effort to define such a mechanism took much longer than expected. (In its current form, that effort is an active [project group](https://github.com/rust-lang/project-ffi-unwind).)

As one of the co-shepherds of the project group, I don't know if the lang team expects or wants me to weigh in on the options listed above, but I do think that the decision cannot rely on the assumption that the opt-out mechanism mentioned above will be ready "soon". The project group does not currently have a design recommendation ready, nor any estimated timeline for delivering such a recommendation. I therefore think it is in the Rust project's best interest that one of the three approaches above be stabilized as soon as possible.

**Comment by nikomatsakis** [comment]
@rfcbot fcp close

I am going to move to close this proposal. I don't think there is much appetite to make unsafe semantically meaningful in this way. I think the main goal should be to produce a final decision on the semantics of the C ABI (and whether it should permit unwinding, or whether some alternate means should be added). We are working towards that goal -- the real question then becomes whether, in the meantime, we wish to remove the UB and -- if so -- how.  I guess we can discuss that particular question in the Zulip, but it seems like most times that we have done so we've reached a sort of stalemate. I'm not sure yet what I think about it -- I guess I think I'd prefer to try to finish up the debate about our overall direction (which seems to me to be close to the point where we can make a call) than to spend energy debating the pros/cons here.

**Comment by joshtriplett** [comment]
@rfcbot reviewed

**Comment by gnzlbg** [comment]
Im fine with closing this, as long as there is a near term alternative in
the very near  future for closing the soundness bug (eg another PR would
do). So as concern I’d like to raise out the issue: what’s the albternative?

AFAIK this PR closes the soundness bug and is compatible with all
alternatives I know of.

I also believe that it would be unfortunate to make the unsafe keyword have
an impact on code generation. But according to this PR, that would only be
temporary until the underlying issues are solved. Those issues do not have
an RFC right now, so AFAICT it might take a couple of years to solve them.
If that were the case, I’d prefer to close the soundness bug now with this
PR, and leave it up to future RFCs not re-open it.

**Comment by gnzlbg** [comment]
Also, this is forward compatible with all soundness fixes to the soundness
bug. So I’d like to hear why should we leave a soundness bug open in safe
Rust when there is a patch to close such but without any kind of perf or
usability drawback for anybody.

What’s the value of leaving programmers using safe Rust code exposed to
undefined behavior when there is a way to protect them from UB?

This PR protects such programmers.

It’s IMO quite hard to argue that we should allow safe Rust to exhibit UB
for consistency with unsafe Rust code. The whole point of using safe Rust
is to not be exposed to UB in the first place.
-- 
Dipl.-Ing. Gonzalo Brito Gadeschi
Institute of Aerodynamics and Chair of Fluid Mechanics
RWTH Aachen University
Wuellnerstraße 5a
D-52062 Aachen
Germany
Phone: ++49-(0)241-80-94821
Fax: ++49-(0)241-80-92257
E-mail:  g.brito@aia.rwth-aachen.de
Internet: www.aia.rwth-aachen.de

**Pull request closed without merging**